pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	obj, found, err = p.cache.GetByKey(image)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	return nil, false, nil
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a provided CRI auth plugin based
+// on it's name provided in CRIAuthPluginConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	args         []string
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary in the following format:
+//
+//  $ <plugin-name> get-credentials arg1=value1 arg2=value2
+//
+// The plugin is expected to receive the CRIAuthPluginRequest API via stdin and is expected
+// to return the CRIAuthPluginResponse via stdout.
+func (e *execPlugin) ExecPlugin(image string) (*criauth.CRIAuthPluginResponse, error) {
+	pluginBin := filepath.Join(e.pluginBinDir, e.name)
+	args := []string{"get-credentials"}
+	args = append(args, e.args...)
+
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
+
+	// Use a catch-all timeout of 1 minute for all exec-based plugins, this should leave enough
+	// head room in case a plugin needs to retry a failed request while ensuring an exec plugin
+	// does not run forever. In the future we may want this timeout to be tweakable from the plugin
+	// config file.
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, pluginBin, args...)
+	cmd.Stdout, cmd.Stderr, cmd.Stdin = stdout, stderr, stdin
+
+	authRequest := &criauth.CRIAuthPluginRequest{Image: image}
+	data, err := encodeRequest(authRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode auth request: %v", err)
+	}
+
+	if _, err := stdin.Write(data); err != nil {
+		return nil, fmt.Errorf("error writing to stdin: %v", err)
+	}
+
+	err = cmd.Run()
+	if ctx.Err() == context.DeadlineExceeded {
+		return nil, fmt.Errorf("timed out execing CRI auth plugin %s for image %s", e.name, image)
+	}
+
+	if err != nil {
+		klog.V(2).Infof("Error execing CRI auth plugin, stderr: %v", stderr.String())
+		return nil, fmt.Errorf("error execing CRI auth plugin %s for image %s: %v", e.name, image, err)
+	}
+
+	response, err := decodeResponse(stdout.Bytes())
+	if err != nil {
+		// err is explicitly not wrapped since it may contain credentials in the response.
+		return nil, errors.New("error decoding CRI auth plugin response from stdout")
+	}
+
+	return response, nil
+}
+
+// encodeRequest encodes the internal CRIAuthPluginRequest type into the v1alpha1 version in json
+func encodeRequest(request *criauth.CRIAuthPluginRequest) ([]byte, error) {
```

Comment:
take the apiVersion requested in the plugin config and encode to the right version here